sdk/javascript/src/nem/MessageEncoder.js

```diff
@@ -0,0 +1,82 @@
+const { deriveSharedKey, deriveSharedKeyDeprecated } = require('./SharedKey');
+const { MessageType, Message } = require('./models');
+const {
+	concatArrays, decodeAesGcm, encodeAesGcm, encodeAesCbc, decodeAesCbc
+} = require('../impl/CipherHelpers');
+
+/**
+ * Encrypts and encodes messages between two parties.
+ */
+class MessageEncoder {
+	/**
+	 * Creates message encoder around key pair.
+	 * @param {KeyPair} keyPair Key pair.
+	 */
+	constructor(keyPair) {
+		this.keyPair = keyPair;
+	}
+
+	/**
+	 * Tries to decode encoded message, returns tuple:
+	 *  * true, message - if message has been decoded and decrypted
+	 *  * false, encodedMessage - otherwise
+	 * @param {PublicKey} recipientPublicKey Recipient's public key.
+	 * @param {Uint8Array} encodedMessage Encoded message
+	 * @returns {array} Tuple containing decoded status and message.
+	 */
+	tryDecode(recipientPublicKey, encodedMessage) {
+		if (MessageType.ENCRYPTED !== encodedMessage.messageType)
+			throw new Error('invalid message format');
+
+		try {
+			const message = decodeAesGcm(deriveSharedKey, this.keyPair, recipientPublicKey, encodedMessage.message);
+			return [true, message];
+		} catch (exception) {
+			if ('Unsupported state or unable to authenticate data' !== exception.message)
+				throw exception;
+		}
+
+		try {
+			const message = decodeAesCbc(deriveSharedKeyDeprecated, this.keyPair, recipientPublicKey, encodedMessage.message);
+			return [true, message];
+		} catch (exception) {
+			if (!exception.message.includes('digital envelope routines:EVP_DecryptFinal_ex:bad decrypt'))
```

Comment:
I think when data itself is malformed (i.e. byte swapped), this will _probably_ result in same throw, not sure if that's a problem